jetty-client/src/main/java/org/eclipse/jetty/client/AbstractConnectionPool.java

```diff
@@ -150,90 +216,167 @@ protected Connection acquire(boolean create)
      * @param maxPending the max desired number of connections scheduled for opening,
      * or a negative number to always trigger the opening of a new connection
      */
-    protected void tryCreate(int maxPending)
+    Future<Void> tryCreate(int maxPending)
     {
-        while (true)
-        {
-            long encoded = connections.get();
-            int pending = AtomicBiInteger.getHi(encoded);
-            int total = AtomicBiInteger.getLo(encoded);
+        CompletableFuture<Void> future = new CompletableFuture<>();
+        int pending = getPendingConnectionCount();
+        int total = pool.size();
+        int maxConnections = pool.getMaxEntries();
 
-            if (LOG.isDebugEnabled())
-                LOG.debug("tryCreate {}/{} connections {}/{} pending", total, maxConnections, pending, maxPending);
+        if (LOG.isDebugEnabled())
+            LOG.debug("tryCreate {}/{} connections {}/{} pending", total, maxConnections, pending, maxPending);
 
-            if (total >= maxConnections)
-                return;
+        if (maxPending >= 0 && pending >= maxPending)
+        {
+            future.complete(null);
+            return future;
+        }
 
-            if (maxPending >= 0 && pending >= maxPending)
-                return;
+        if (LOG.isDebugEnabled())
+            LOG.debug("newConnection {}/{} connections {}/{} pending", total + 1, maxConnections, pending + 1, maxPending);
 
-            if (connections.compareAndSet(encoded, pending + 1, total + 1))
+        Pool<Connection>.Entry entry = pool.addReservation();
```

Comment:
I think that it is strange to try to enforce the pending externally to the pool (which calls pending reservations).
Why not do `pool.addReservation(maxPending)` ?

Reply:
Yes, this makes sense. Done.